### DISCUSSION TITLE
generic Assets

### DISCUSSION THREAD

**Discussion by maryamznegui** [dagster-io/dagster#16358] [Q&A]
Hello 

I'm trying to create a two generic assets: 
asset 1 : to perform any get request based on configuration file 
asset 2 : to push data to s3 and takes as parameter the output of asset 1

i have configuration file where i define the assets named assets.json 
contains data like below 
```python
assets =  
 [ {
        "asset-name": "xxxx",
         "s3_operation_upstream": "get_operation_xxxx",
        "s3-bucket": "xxxx-bucket",
        "endpoint": "xxxx",
        "secret-key": "xxxx-key",
        "s3-path": "test/xxx/xxx",
    },

  {
        "asset-name": "yyyyy",
         "s3_operation_upstream": "get_operation_yyyy"
        "s3-bucket": "yyyyy-bucket",
        "endpoint": "yyyyy",
        "secret-key": "yyyy-key",
        "s3-path": "test/yyy/yyy",
    }
]
```
and this is the function the creates the generic assets 

```python
def build_asset( entry) -> [AssetsDefinition, AssetsDefinition]:

    @asset(
        key=AssetKey.from_user_string(f"get_operation_{entry['asset-name']}"),
        auto_materialize_policy=AutoMaterializePolicy.eager())
    def asset_template_get():
        return get_operation(entry)

    @asset(
        key=AssetKey(f"s3_operation_{entry['asset-name']}"),
        deps=[AssetKey(entry['s3_operation_upstream'])],
        ins={"upstream": AssetIn(entry['s3_operation_upstream'])},
        auto_materialize_policy=AutoMaterializePolicy.eager()
    )
    def asset_template_s3(upstream):
        s3_operation(entry, upstream)

        return asset_template_s3

    return [asset_template_get, asset_template_s3]
```
then adding the assets to list of AssetsDefinition: 

```python
data = load_json("assets.json")
factory_assets = [build_asset(entry)[0] for entry in data['assets']]
factory_assets.extend(build_asset(entry)[1] for entry in data['assets'])
```

i can see in the UI the 4 assets and every get asset linked to s3 asset , get assets are running fine however i'm not able to run s3 assets im getting below error 

```
__ASSET_JOB_0 cannot be executed with the provided config. Please fix the following errors:

Missing required config entry "ops" at the root. Sample config for missing entry: {'ops': {'s3_operation_xxxx': {'inputs': {'upstream': '<selector>'}}}}
```
Appreciate your thoughts on this, and please let me know if there is a better way to implement the generic assets with creating the dependency between them 

Many Thanks!

**Comment by tom-disco** [comment]
Hey, I'm also working through something similar to this. Best I've found so far is splitting the factory function into two sections, and have a bool parameter (like `generic`) that decides on the logic:
```py
def generate_assets(config: MyConfig, generic: bool = False):

    if generic:

        @asset(name=asset_name)
        def upstream_asset(
            context: OpExecutionContext,
            config: MyConfig,
        ) -> str:
            """
            Generate a UUID asset to pass to other assets to track a single job
            """
            u = str(uuid.uuid4())
            context.log.info(f"Generated UUID {u} for Config {config.__dict__}")
            return u

    else:

        @asset(name=asset_name)
        def upstream_asset(
            context: OpExecutionContext,
        ) -> str:
            """
            Generate a UUID asset to pass to other assets to track a single job
            """
            u = str(uuid.uuid4())
            context.log.info(f"Generated UUID {u} for Config {config.__dict__}")
            return u
  ...
```

So the `generic` param is just dictating whether to use a passed in config or to force it to be entered at runtime
I _think_ this can be extrapolated by returning the config as an asset and then passing it to other assets so that the `if/then` only needs to apply to the first asset, then the rest just use whatever config happens to have been used by the first one

**Reply by tom-disco** [reply]
I also think this is a hideous way to handle this pattern

**Reply by AlejandroUPC** [reply]
Have been facing somewhat a similar way to handle the dynamic creation of assets, feels a bit odd so I agree. Creating dynamic ops seems so much easier.

**Reply by sryza** [reply]
@tom-disco from what I can tell, the difficulty you're hitting looks different than the issue reported at the top of this discussion (where no explicit Config is provided). Would you mind creating a separate discussion for it?

**Comment by sryza** [comment]
Hi @maryamznegui - one thing that sticks out to me about your code example is that you're using both `deps` and `ins`.  In general, you should only use one. @jamiedemaria should we be enforcing this?

The second thing that sticks out is that `ins={"upstream": AssetIn(entry['s3_operation_upstream'])}` doesn't look like it matches `key=AssetKey.from_user_string(f"get_operation_{entry['asset-name']}"),` If you want the output of the first asset to go into the input of the second asset, I'd expect that these should be the same.
